src/qibo/noise_model.py

```diff
@@ -0,0 +1,398 @@
+        "t2": tuple(parameters[qubits : 2 * qubits]),
+        "gate_time": tuple(parameters[2 * qubits : 2 * qubits + 2]),
+        "excited_population": 0,
+        "depolarizing_error": tuple(parameters[2 * qubits + 2 : 2 * qubits + 4]),
+        "bitflips_error": (
+            parameters[2 * qubits + 4 : 3 * qubits + 4],
+            parameters[3 * qubits + 4 : 4 * qubits + 4],
+        ),
+        "idle_qubits": idle_qubits,
+    }
+
+    noisy_circ = noisy_circuit(circuit, params)
+    freq = backend.execute_circuit(circuit=noisy_circ, nshots=nshots).frequencies()
+    prob = freq_to_prob(freq)
+
+    hellinger_fid = hellinger_fidelity(target_prob, prob)
+
+    if error == True:
+        return [hellinger_fid, hellinger_shot_error(target_prob, prob, nshots)]
+    else:
+        return hellinger_fid
+
+
+class CompositeNoiseModel:
+    """Class associated with a realistic representation of a noisy circuit modeled by the :func:`qibo.noise_model.noisy_circuit`.
+    This class is able to fit the parameters of the noise model to reproduce an experimental realization of the circuit
+    you want to simulate.
+
+    Args:
+        noisy_circuit (qibo.models.Circuit): the noisy circuit. See :func:`qibo.noise_model.noisy_circuit`.
+        params (dictionary): the parameters of the noise model. See :func:`qibo.noise_model.noisy_circuit`.
+        hellinger (float): current value of the hellinger fidelity between the noisy simulation and the given target result.
+        hellinger0 (dictionary): the fidelity and the shot error fidelity  organized as {"fidelity": (float) f, "shot_error": (float) e}.
+    """
+
+    def __init__(self, params):
+        self.noisy_circuit = {}
+        self.params = params
+        self.hellinger = {}
+        self.hellinger0 = {}
+
+    def apply(self, circuit):
+        """Creates the noisy circuit from the circuit given as argument by using the :func:`qibo.noise_model.noisy_circuit`.
+        Args:
+            circuit (qibo.models.Circuit): the circuit you want to simulate.
+        """
+        self.noisy_circuit = noisy_circuit(circuit, self.params)
+
+    def fit(
+        self,
+        target_result,
+        bounds=True,
+        f_min_rtol=None,
+        backend=None,
+    ):
+        r"""Performes the fitting procedure of the noise model parameters, using the method nlopt.opt from the library nplot. The fitting procedure is implemented to maximize the hellinger fidelity calculated using the :func:`qibo.noise_model.loss` between the probability distribution function estimated by the noise model and the one measured experimentally. Since, we are using probability distribution functions estimated using a finite number of shots, the hellinger fidelity is going to have an error caused by an imperfect estimation of the probabilities. This method takes into account this effect and stops when the fidelity reaches a corrected maximum $1-\epsilon$, with $\epsilon$=:func:`qibo.noise_model.hellinger_shot_error`.
```

Comment:
```suggestion
        r"""Performs the fitting procedure of the noise model parameters, using the method `nlopt.opt` from the library NLopt. The fitting procedure is implemented to maximize the hellinger fidelity calculated using the :func:`qibo.noise_model.loss` between the probability distribution function estimated by the noise model and the one measured experimentally. Since we are using probability distribution functions estimated using a finite number of shots, the hellinger fidelity is going to have an error caused by an imperfect estimation of the probabilities. This method takes into account this effect and stops when the fidelity reaches a corrected maximum $1-\epsilon$, with $\epsilon$=:func:`qibo.noise_model.hellinger_shot_error`.
```